contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,469 @@
+import {PoolManager} from "@uniswap/v4-core/contracts/PoolManager.sol";
+import {Pool} from "@uniswap/v4-core/contracts/libraries/Pool.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {BaseHook} from "../BaseHook.sol";
+
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {CurrencyLibrary, Currency} from "@uniswap/v4-core/contracts/libraries/CurrencyLibrary.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {BalanceDelta, BalanceDeltaLibrary} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {IERC20Minimal} from "@uniswap/v4-core/contracts/interfaces/external/IERC20Minimal.sol";
+import {ILockCallback} from "@uniswap/v4-core/contracts/interfaces/callback/ILockCallback.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/libraries/PoolId.sol";
+import {FullMath} from "@uniswap/v4-core/contracts/libraries/FullMath.sol";
+import {UniswapV4ERC20} from "./UniswapV4ERC20.sol";
+import {IUniswapV4ERC20} from "./IUniswapV4ERC20.sol";
+import {SafeMath} from "./SafeMath.sol";
+import {Math} from "./Math.sol";
+import {Position} from "@uniswap/v4-core/contracts/libraries/Position.sol";
+import "@uniswap/v4-core/contracts/libraries/FixedPoint128.sol";
+
+import "forge-std/console.sol";
+
+import "../libraries/LiquidityAmounts.sol";
+
+contract FullRange is BaseHook {
+    using CurrencyLibrary for Currency;
+    using PoolIdLibrary for IPoolManager.PoolKey;
+
+    /// @notice Thrown when trying to interact with a non-initialized pool
+    error PoolNotInitialized();
+
+    /// @dev Min tick for full range with tick spacing of 60
+    int24 internal constant MIN_TICK = -887220;
+    /// @dev Max tick for full range with tick spacing of 60
+    int24 internal constant MAX_TICK = -MIN_TICK;
+
+    uint256 public constant MINIMUM_LIQUIDITY = 10 ** 3;
+
+    uint256 internal blockNumber;
+
+    struct CallbackData {
+        address sender;
+        IPoolManager.PoolKey key;
+        IPoolManager.ModifyPositionParams params;
+        bool rebalance;
+    }
+
+    struct HookPosition {
+        // the tick range of the position
+        // int24 tickLower;
+        // int24 tickUpper;
+        // the liquidity of the position
+        uint128 liquidity;
+        // the fee growth of the aggregate position as of the last action on the individual position
+        uint256 feeGrowthInside0LastX128;
+        uint256 feeGrowthInside1LastX128;
+        // how many uncollected tokens are owed to the position, as of the last computation
+        uint128 tokensOwed0;
+        uint128 tokensOwed1;
+    }
+
+    mapping(PoolId => address) public poolToERC20;
+    mapping(PoolId => HookPosition) public poolToHookPosition;
+
+    constructor(IPoolManager _poolManager) BaseHook(_poolManager) {
+        blockNumber = block.number; // TODO: potentially set this on before initialization?
```

Comment:
yes, `block.number` will have to be per pool

Reply:
might be worth a test testing that two diff pools get updated on first pool action of the respective pool. having one block.number for all pools would make that test fail.